relay/pkg/block/forwarder.go

```diff
@@ -56,20 +63,126 @@ func RunForwarder(
 		hostChain:    hostChain,
 		headersQueue: make(chan *btc.Header, headersQueueSize),
 		errChan:      make(chan error, 1),
+		quit:         make(chan bool, 1),
 	}
 
+	go forwarder.pullingLoop(ctx)
 	go forwarder.pushingLoop(ctx)
 
 	return forwarder
 }
 
+func (f *Forwarder) findBestBlock() (*btc.Header, error) {
+	currentBestDigest, err := f.hostChain.GetBestKnownDigest()
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Infof("best known digest returned from Ethereum network: %s",
+		hex.EncodeToString(currentBestDigest[:]),
+	)
+
+	bestHeader, err := f.btcChain.GetHeaderByDigest(currentBestDigest)
+	if err != nil {
+		return nil, err
+	}
+
+	betterOrSameHeader, err := f.btcChain.GetHeaderByHeight(bestHeader.Height)
+	if err != nil {
+		return nil, err
+	}
+
+	// see if there's a better block at that height
+	// if so, crawl backwards
+
+	// TODO: Is it ever possible that bestHeader and betterOrSameHeader are not
+	// equal?
+	// TODO: Consider just comparing hashes - it should be enough
+	for !headersEqual(bestHeader, betterOrSameHeader) {
+		bestHeader, err = f.btcChain.GetHeaderByDigest(bestHeader.PrevHash)
+		if err != nil {
+			return nil, err
+		}
+
+		betterOrSameHeader, err = f.btcChain.GetHeaderByHeight(bestHeader.Height)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return bestHeader, nil
+}
+
+func (f *Forwarder) pullingLoop(ctx context.Context) {
+	logger.Infof("running forwarder pulling loop")
+
+	latestHeader, err := f.findBestBlock()
+	if err != nil {
+		f.errChan <- fmt.Errorf(
+			"failure while trying to find best block for pulling loop: [%v]",
+			err,
+		)
+		return
+	}
+
+	logger.Infof("starting pulling loop with header: hash %s at height %d",
+		latestHeader.Hash.String(), latestHeader.Height)
+
+	latestHeight := latestHeader.Height + 1
+	lastAdded := &btc.Header{}
+
+	for {
+		select {
+		case <-ctx.Done():
+			logger.Infof("forwarder context is done")
+			return
+		case <-f.quit:
+			return
+		default:
```

Comment:
I created `pull.go` file and moved there `findBestBlock` and also functions `pullHeaderFromBtcNetwork` and `pushHeaderToQueue` that I created.

As for extracting more logic from `default`, it is hard to extract anything more, as there is a loop there and the logic in that loop uses variables (`nextHeaderHeight` and `lastAdded`) which are outside the scope of the loop.

I would have to pass them to the extracted function.

Reply:
What about making `nextHeaderHeight` and `lastAdded` private fields of the `Forwarder` struct? We already use `processedHeaders` on the push side so we may introduce:
- `nextPulledHeaderHeight`
- `lastPulledHeader`
This may open more possibilities to take processing logic out of the pulling loop. What do you think?

Reply:
I created  `pullNextHeader` and moved it to `pull.go`.
I could move there even more logic from `default` there, but I think I will be easier to think about the pulling loop if we leave the sleeping part on top, so it is clearly visible.

Reply:
I get your point about the sleeping part but I think that sleep is strictly related to the "pull from BTC chain" part not the entire pull loop. My proposition of how the `default` case should look:
```
		default:
                        logger.Infof("pulling new header from BTC chain")

			header, err := f.pullHeaderFromBtcChain(ctx)
			if err != nil {
				f.errChan <- fmt.Errorf("could not pull header: [%v]", err)
				return
			}

                        logger.Infof("pushing new header to the queue")

			f.pushHeaderToQueue(header)
```
And that's it. Regarding methods used here look at the next comment.